<a href="https://colab.research.google.com/github/Berte08/CSI4106_A3/blob/main/DEVOIR_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **DEVOIR** 4 - Classification de textes

1. Numéro du groupe: 31 \\
   Noms des membres: TATA SAIDATOU BERTE \\
   Numéros d'étudiants des membres: 300100935 \\
   Titre: Devoir4 \\


**L'objectif** de ce devoir est d'analyser et de comparer les performances de différents modèles de machine learning, spécifiquement la régression logistique et le Perceptron multicouches (MLP), sur un ensemble de données multi-classes.

**IMPORTATION DES BIBLIOTHEQUES IMPORTANTES ET ACCES DES DONNEES**

In [1]:
!pip install spacy

In [2]:
!python -m spacy download en_core_web_sm

2023-12-02 20:16:43.648272: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-12-02 20:16:43.648340: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-12-02 20:16:43.648380: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-12-02 20:16:43.663646: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-12-02 20:16:46.105121: W tensorflow/compiler/

In [3]:
import spacy
import pandas as pd
import numpy as np

# **1-choix des ensembles de données textuelles**

 La base de données choisie est  CNN theme classification

In [4]:


url ="https://github.com/Berte08/CSI4106_A3/raw/main/reduced_file_cnnnews.csv"

#url ="https://github.com/Berte08/CSI4106_A3/raw/main/reduced_file_AirPassengerReviews.csv"


Chargement de données

In [5]:
print(url)
data = pd.read_csv(url)

https://github.com/Berte08/CSI4106_A3/raw/main/reduced_file_cnnnews.csv


In [6]:
data.head()

ID                                              Title  \
0   1  Candy factory didn't evacuate concerned worker...   
1   2  Baltimore police ask for public's help identif...   
2   3  An arrest warrant has been issued for a suspec...   
3   4  115 improperly stored human remains found in C...   
4   5  Bronx day care provider and 2 others indicted ...   

                                         Description  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                                Body  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                            Keywords Theme  \
0  accident investigations, accidents, accidents,...    us   
1  baltimore, brand safety-nsf crime, brand safet...    us   
2  arrest warrants, arrests, brand safety-nsf cri...    us   
3  brand safety-nsf death, brand safety-nsf sensi...    us   
4  brand safety-nsf crime, brand safety-nsf death...    us   

                                                Link  
0  https://edition.cnn.com/2023/10/06/us/pennsylv...  
1  https://edition.cnn.com/2023/10/06/us/morgan-s...  
2  https://edition.cnn.com/2023/10/06/us/josh-kru...  
3  https://edition.cnn.com/2023/10/06/us/colorado...  
4  https://edition.cnn.com/2023/10/05/us/bronx-da...

# **2. Effectuez des tâches de la pipeline TAL sur des données textuelle**

Chargement du modele spacy

In [7]:
nlp = spacy.load("en_core_web_sm")

L'application du pipeline à chaque phrase crée un document dans lequel chaque mot est un objet Token.

Doc: https://spacy.io/api/doc

Token: https://spacy.io/api/token

In [8]:
#Apply nlp pipeline to the column that has your sentences (the text that will serve as input features).
data['tokenized'] = data['Body'].apply(nlp)

In [9]:
data.head()

ID                                              Title  \
0   1  Candy factory didn't evacuate concerned worker...   
1   2  Baltimore police ask for public's help identif...   
2   3  An arrest warrant has been issued for a suspec...   
3   4  115 improperly stored human remains found in C...   
4   5  Bronx day care provider and 2 others indicted ...   

                                         Description  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                                Body  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                            Keywords Theme  \
0  accident investigations, accidents, accidents,...    us   
1  baltimore, brand safety-nsf crime, brand safet...    us   
2  arrest warrants, arrests, brand safety-nsf cri...    us   
3  brand safety-nsf death, brand safety-nsf sensi...    us   
4  brand safety-nsf crime, brand safety-nsf death...    us   

                                                Link  \
0  https://edition.cnn.com/2023/10/06/us/pennsylv...   
1  https://edition.cnn.com/2023/10/06/us/morgan-s...   
2  https://edition.cnn.com/2023/10/06/us/josh-kru...   
3  https://edition.cnn.com/2023/10/06/us/colorado...   
4  https://edition.cnn.com/2023/10/05/us/bronx-da...   

                                           tokenized  
0  (An, eastern, Pennsylvania, candy, factory, di...  
1  (Two, shooters, were, involved, in, an, attack...  
2  (Authorities, in, Pennsylvania, say, they, hav...  
3  (An, investigation, into, more, than, 115, bod...  
4  (A, Bronx, day, care, provider, ,, her, husban...

Un token a plusieurs attributs, tel part-of-speech (pos_), lemma (lemma_), etc. Regardez la documentation pour voir l'ensemble des attributs.

La fonction suivante est un exemple de la façon dont vous pouvez récupérer les parties du discours (POS)à partir d'une phrase. Nous renvoyons la lemmatisation car nous voulons uniquement le mot à l'infinitif.

In [10]:
#create empty dataframes that will store your derived datasets


derived_dataset1 = pd.DataFrame(columns=['ID', 'pos'])
derived_dataset2 = pd.DataFrame(columns=['ID', 'pos-np'])

In [11]:
# Fonction pour extraire des POS spécifiques et lemmatiser
def get_pos(sentence, wanted_pos):
    words = []
    for token in sentence:
        if token.pos_ in wanted_pos:
          #  lemma pour obtenir la forme de base du mot
            words.append(token.lemma_)
    return ' '.join(words)

In [12]:
#As an example, we use the above function to fetch all the verbs. We store this information in our first derived dataset
derived_dataset1['pos'] = data['tokenized'].apply(lambda sent : get_pos(sent, ['VERB']))
derived_dataset1['ID'] = data['ID']

In [13]:
derived_dataset1.head()

ID                                                pos
0   1  evacuate say smell leave injure announce base ...
1   2  involve injure confirm ask identify see walk c...
2   3  say issue suspect kill advocate identify consi...
3   4  find store offer take say find go say have nee...
4   5  indict expose include die say indict accord sa...

Dans notre Derived-Dataset-1, on a décidé de regarder de près les verbes dans les articles de CNN.Pour les raison suivantes:

Les verbes sont importants dans les articles , ils racontent ce qui se passe. Ils sont au le cœur des histoires - ils montrent les actions . En prenant les verbes, on saisit les parties les plus actives des articles, ce qui nous aide à mieux comprendre l'histoire.

 Lemmatisatiser les verbes nous aide à regrouper les différents formes d'un même verbe. Par exemple, "running", "ran", et "run" sont tous regroupés sous "run". C'est utile pour repérer les sujets principaux .

In [14]:
#For Derived Dataset 2, you also need to include Named Entities
#Below is just an example of obtaining such entities on a specific sentence, but you would do NER
#on the dataset of your choice.
#You can choose the types of entities (dates, organization, people) that you want,
#and then in your derived dataset, just make sure you include these entities separated by spaces (as shown for verbs)
#in a previous cell.

def extract_entities_and_np(sentence):
    entities_and_np = []
    # Extraction des entités
    for ent in sentence.ents:
        entities_and_np.append(ent.text)
    # Extraction des noms propres
    for token in sentence:
        if token.pos_ == 'PROPN':
            entities_and_np.append(token.text)
    return ' '.join(entities_and_np)

In [15]:
# Appliquer la fonction à chaque texte tokenisé
derived_dataset2['pos-np'] = data['tokenized'].apply(extract_entities_and_np)
derived_dataset2['ID'] = data['ID'] # Inclure l'ID pour référence

In [16]:
# Afficher les premières lignes de Derived-Dataset-2
print(derived_dataset2.head())

   ID                                             pos-np
0   1  Pennsylvania March seven nearly a dozen the US...
1   2  Two Morgan State University Baltimore this wee...
2   3  Pennsylvania Josh Kruger earlier this week Rob...
3   4  more than 115 Colorado months Penrose about 30...
4   5  Bronx four 1-year-old last month Bronx Thursda...


Maintenant que vous disposez de vos ensembles de données dérivés, vous pouvez effectuer votre tâche de classification.

# **3. Effectuez une étude empirique de classification
**


a-Features

In [39]:
# TF-IDF pour le dataset original
vectorizer_original = TfidfVectorizer()
X_original = vectorizer_original.fit_transform(data['Body'])  # Assurez-vous que 'Body' est la colonne contenant le texte.
X_derived1 = X_bow_stopwords_1
X_derived2 = X_tfidf_2
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(data['Theme'])



In [17]:

print("Derived Dataset 1:")
print(derived_dataset1.head())


print("\nDerived Dataset 2:")
print(derived_dataset2.head())


Derived Dataset 1:
   ID                                                pos
0   1  evacuate say smell leave injure announce base ...
1   2  involve injure confirm ask identify see walk c...
2   3  say issue suspect kill advocate identify consi...
3   4  find store offer take say find go say have nee...
4   5  indict expose include die say indict accord sa...

Derived Dataset 2:
   ID                                             pos-np
0   1  Pennsylvania March seven nearly a dozen the US...
1   2  Two Morgan State University Baltimore this wee...
2   3  Pennsylvania Josh Kruger earlier this week Rob...
3   4  more than 115 Colorado months Penrose about 30...
4   5  Bronx four 1-year-old last month Bronx Thursda...


b- Bag-of-Words

Cette étape transforme les textes de derived_dataset1 et derived_dataset2 en une représentation numérique utilisant la méthode Bag-of-Words.

In [18]:
from sklearn.feature_extraction.text import CountVectorizer

# Bag-of-Words pour derived_dataset1
vectorizer_bow_1 = CountVectorizer()
X_bow_1 = vectorizer_bow_1.fit_transform(derived_dataset1['pos'])

# Bag-of-Words pour derived_dataset2
vectorizer_bow_2 = CountVectorizer()
X_bow_2 = vectorizer_bow_2.fit_transform(derived_dataset2['pos-np'])


In [19]:

print("Forme des Features BoW pour Derived Dataset 1:", X_bow_1.shape)
print("Forme des Features BoW pour Derived Dataset 2:", X_bow_2.shape)


Forme des Features BoW pour Derived Dataset 1: (1412, 4191)
Forme des Features BoW pour Derived Dataset 2: (1412, 15767)


c-Stop Words

Suppression des mots vides pour réduire le bruit dans les données

In [20]:
# Bag-of-Words avec suppression des mots vides pour derived_dataset1
vectorizer_bow_stopwords_1 = CountVectorizer(stop_words='english')
X_bow_stopwords_1 = vectorizer_bow_stopwords_1.fit_transform(derived_dataset1['pos'])

# Bag-of-Words avec suppression des mots vides pour derived_dataset2
vectorizer_bow_stopwords_2 = CountVectorizer(stop_words='english')
X_bow_stopwords_2 = vectorizer_bow_stopwords_2.fit_transform(derived_dataset2['pos-np'])


In [21]:

print("Forme des Features BoW  pour Derived Dataset 1:", X_bow_stopwords_1.shape)
print("Forme des Features BoW  pour Derived Dataset 2:", X_bow_stopwords_2.shape)


Forme des Features BoW  pour Derived Dataset 1: (1412, 4143)
Forme des Features BoW  pour Derived Dataset 2: (1412, 15579)


d-TF-IDF

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF pour derived_dataset1
vectorizer_tfidf_1 = TfidfVectorizer()
X_tfidf_1 = vectorizer_tfidf_1.fit_transform(derived_dataset1['pos'])

# TF-IDF pour derived_dataset2
vectorizer_tfidf_2 = TfidfVectorizer()
X_tfidf_2 = vectorizer_tfidf_2.fit_transform(derived_dataset2['pos-np'])


In [23]:
print("Forme des Features TF-IDF pour Derived Dataset 1:", X_tfidf_1.shape)
print("Forme des Features TF-IDF pour Derived Dataset 2:", X_tfidf_2.shape)


Forme des Features TF-IDF pour Derived Dataset 1: (1412, 4191)
Forme des Features TF-IDF pour Derived Dataset 2: (1412, 15767)


2-Les modeles

**Module** pour la modelisation

In [41]:
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, precision_recall_fscore_support


In [33]:
# Encodage des labels
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(data['Theme'])

# Séparation en ensembles d'entraînement et de test
X_train, X_test, y_train, y_test = train_test_split(X_bow_1, y_encoded, test_size=0.2, random_state=42)


a-Modèle de régression logistique

In [57]:
logistic_model = LogisticRegression(max_iter=500, solver='lbfgs', random_state=42)


b-. Modèle Perceptron multicouches

In [58]:
mlp_model = MLPClassifier(hidden_layer_sizes=(100,), max_iter=500, activation='relu', solver='adam', random_state=42)


# **3-Entraînez/testez/évaluez vos 2 modèles**

 a-Validation croisée

In [59]:
kf = KFold(n_splits=4, shuffle=True, random_state=42)

B-Fonction d'évaluation

In [60]:
from sklearn.metrics import classification_report
# Fonction pour évaluer un modèle
def evaluate_model(model, X, y):
    scores = cross_val_score(model, X, y, cv=kf, scoring='accuracy')
    print(f"Scores de validation croisée: {scores}")
    print(f"Moyenne des scores: {scores.mean()}")

    # Entraînement et évaluation
    model.fit(X, y)
    y_pred = model.predict(X)
    print(classification_report(y, y_pred))

# Évaluation pour chaque ensemble de données
for X, name in [(X_original, "Original"), (X_derived1, "Derived1"), (X_derived2, "Derived2")]:
    print(f"--- Évaluation sur {name} ---")
    print("Régression Logistique:")
    evaluate_model(logistic_model, X, y_encoded)

    print("Perceptron Multicouches:")
    evaluate_model(mlp_model, X, y_encoded)

--- Évaluation sur Original ---
Régression Logistique:
Scores de validation croisée: [0.73654391 0.76487252 0.77620397 0.76203966]
Moyenne des scores: 0.7599150141643061


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           0       0.94      0.97      0.95       201
           1       0.00      0.00      0.00         1
           2       0.94      1.00      0.97       113
           3       0.94      0.99      0.96       139
           4       0.92      0.71      0.80       126
           5       0.92      0.91      0.92       174
           6       0.99      0.98      0.99       152
           7       0.00      0.00      0.00        15
           8       0.00      0.00      0.00        11
           9       0.98      0.73      0.84        84
          10       0.97      1.00      0.98        97
          11       0.83      0.99      0.90       299

    accuracy                           0.92      1412
   macro avg       0.70      0.69      0.69      1412
weighted avg       0.90      0.92      0.91      1412

Perceptron Multicouches:
Scores de validation croisée: [0.81303116 0.8611898  0.8611898  0.83852691]
Moyenne des scores: 0.8434844192

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Scores de validation croisée: [0.70254958 0.67705382 0.71671388 0.65155807]
Moyenne des scores: 0.6869688385269122


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


              precision    recall  f1-score   support

           0       1.00      1.00      1.00       201
           1       1.00      1.00      1.00         1
           2       1.00      1.00      1.00       113
           3       1.00      1.00      1.00       139
           4       1.00      1.00      1.00       126
           5       1.00      1.00      1.00       174
           6       1.00      1.00      1.00       152
           7       1.00      1.00      1.00        15
           8       1.00      1.00      1.00        11
           9       1.00      1.00      1.00        84
          10       1.00      1.00      1.00        97
          11       1.00      1.00      1.00       299

    accuracy                           1.00      1412
   macro avg       1.00      1.00      1.00      1412
weighted avg       1.00      1.00      1.00      1412

Perceptron Multicouches:
Scores de validation croisée: [0.69688385 0.71104816 0.75070822 0.67988669]
Moyenne des scores: 0.7096317280

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           0       0.95      0.97      0.96       201
           1       0.00      0.00      0.00         1
           2       0.95      0.99      0.97       113
           3       0.97      0.99      0.98       139
           4       0.92      0.57      0.71       126
           5       0.88      0.92      0.90       174
           6       0.99      0.97      0.98       152
           7       0.00      0.00      0.00        15
           8       0.00      0.00      0.00        11
           9       0.99      0.95      0.97        84
          10       0.98      0.99      0.98        97
          11       0.82      0.99      0.90       299

    accuracy                           0.92      1412
   macro avg       0.70      0.70      0.70      1412
weighted avg       0.90      0.92      0.91      1412

Perceptron Multicouches:
Scores de validation croisée: [0.74220963 0.78186969 0.83002833 0.77053824]
Moyenne des scores: 0.7811614730

Affichage de la distribution des classe

In [61]:
print(data['Theme'].value_counts())



world            299
business         201
politics         174
sport            152
health           139
opinions         126
entertainment    113
weather           97
us                84
style             15
travel            11
cnn                1
Name: Theme, dtype: int64


**L'analyse des résultats** montre que le modèle de régression logistique fonctionne bien en général mais a du mal avec les classes qui apparaissent moins souvent. Pour ces classes, il n'est pas très précis. D'un autre côté, le modèle Perceptron Multicouches réussit parfaitement sur toutes les classes. Enfin, certaines classes sont beaucoup plus fréquentes que d'autres dans les données,elles sont desiquilibrer. Cela peut poser problème car les modèles peuvent ne pas apprendre correctement à reconnaître les classes qui sont moins nombreuses.

# **4-Modification de certains paramètres**

Nous allons  changer la taille des couches cachées et la fonction d'activation

Changer la taille des couches cachées

In [65]:
# Augmentation de la taille de la couche cachée
mlp_model_large = MLPClassifier(hidden_layer_sizes=(150,), max_iter=200, activation='relu', solver='adam', random_state=42)
evaluate_model(mlp_model_large, X_original, y_encoded)
evaluate_model(mlp_model_large, X_derived1, y_encoded)
evaluate_model(mlp_model_large, X_derived2, y_encoded)

# Diminution de la taille de la couche cachée
mlp_model_small = MLPClassifier(hidden_layer_sizes=(50,), max_iter=200, activation='relu', solver='adam', random_state=42)
evaluate_model(mlp_model_small, X_original, y_encoded)
evaluate_model(mlp_model_small, X_derived1, y_encoded)
evaluate_model(mlp_model_small, X_derived2, y_encoded)


Scores de validation croisée: [0.81303116 0.85835694 0.86685552 0.84419263]
Moyenne des scores: 0.8456090651558074
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       201
           1       1.00      1.00      1.00         1
           2       1.00      1.00      1.00       113
           3       1.00      1.00      1.00       139
           4       1.00      1.00      1.00       126
           5       1.00      1.00      1.00       174
           6       1.00      1.00      1.00       152
           7       1.00      1.00      1.00        15
           8       1.00      1.00      1.00        11
           9       1.00      1.00      1.00        84
          10       1.00      1.00      1.00        97
          11       1.00      1.00      1.00       299

    accuracy                           1.00      1412
   macro avg       1.00      1.00      1.00      1412
weighted avg       1.00      1.00      1.00      1412

Scores de validati

Changer la Fonction d'Activation

In [66]:
# Utilisation de la fonction d'activation tanh
mlp_model_tanh = MLPClassifier(hidden_layer_sizes=(100,), max_iter=200, activation='tanh', solver='adam', random_state=42)
evaluate_model(mlp_model_tanh, X_original, y_encoded)
evaluate_model(mlp_model_tanh, X_derived1, y_encoded)
evaluate_model(mlp_model_tanh, X_derived2, y_encoded)


Scores de validation croisée: [0.81586402 0.86402266 0.86402266 0.83569405]
Moyenne des scores: 0.8449008498583569
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       201
           1       1.00      1.00      1.00         1
           2       1.00      1.00      1.00       113
           3       1.00      1.00      1.00       139
           4       1.00      1.00      1.00       126
           5       1.00      1.00      1.00       174
           6       1.00      1.00      1.00       152
           7       1.00      1.00      1.00        15
           8       1.00      1.00      1.00        11
           9       1.00      1.00      1.00        84
          10       1.00      1.00      1.00        97
          11       1.00      1.00      1.00       299

    accuracy                           1.00      1412
   macro avg       1.00      1.00      1.00      1412
weighted avg       1.00      1.00      1.00      1412

Scores de validati

Analyse

Quand on a augmenté la taille des couches cachées à 150 neurones, le modèle a montré une petite amélioration. Cela veut dire que le modèle a un peu mieux appris à partir des données, mais la différence n'est pas très grande.

Ensuite, quand on a réduit la taille des couches cachées à 50 neurones, les performances du modèle ont un peu baissé. Cela suggère que le modèle avait un peu plus de mal à comprendre les détails des données, mais l'effet n'était pas très important.

Enfin, en changeant la fonction d'activation pour 'tanh', on n'a pas vu de grands changements par rapport à l'utilisation de 'relu'. Cela signifie que, pour ces données, le type de fonction d'activation n'a pas beaucoup d'importance pour la performance du modèle.

5-Analysez les résultats obtenus

a.
Régression logistique et perceptron multicouches sur différents jeux de données :

Sur l'ensemble original, la régression logistique a montré une performance moyenne avec des scores de validation croisée autour de 0.76. Cependant, sa précision et son rappel étaient faibles pour certaines classes moins fréquentes.
Le modèle perceptron multicouches a démontré une précision et un rappel parfaits sur l'ensemble original, indiquant potentiellement un surajustement.
modifications de la taille des couches cachées :
En augmentant la taille des couches cachées à 150 neurones, une légère amélioration a été observée, mais la différence était minime.
En réduisant la taille des couches cachées à 50 neurones, une légère baisse de la performance a été notée, ce qui peut indiquer une capacité réduite du modèle à capturer la complexité des données.
Changement de la Fonction d'Activation (tanh) :
L’utilisation de la fonction d'activation tanh n'a pas montré de changement significatif par rapport à l'utilisation de la fonction relu.

c-La régression logistique montre une grande variabilité dans ses performances en termes de précision et de rappel selon les classes, particulièrement faible pour les classes moins représentées. Cela pourrait être dû à l'inégalité dans la distribution des classes.
Le modèle Perceptron Multicouches, en revanche, présente une précision et un rappel constants et parfaits pour toutes les classes, ce qui est inhabituel et suggère un surajustement, surtout dans le contexte d'un jeu de données avec une distribution de classe inégale.
